### PR TITLE
FIX: Handle Redis failure during jobs

### PR DESCRIPTION
**Pull request by davidtaylorhq**
MiniScheduler was previously able to detect 'orphaned' jobs when the entire `MiniScheduler::Manager` failed (e.g. a process restart). However, if something went wrong with the Redis connection during a job, the worker thread would be unable to remove the "RUNNING" status, and the job would be stuck in that state until its next scheduled run.

This commit makes a few changes to fix this:
- give each thread its own id, and use an expiring redis key to track its lifetime
- when changing a job to the 'RUNNING' state, update its `current_owner` from the manager's id to the worker thread's id. This ensures that the existing orphan-detection logic will pick up the job if the thread dies.
- ensure worker threads die when there is an internal mini_scheduler exception which could cause inconsistent job state
- automatically start new worker threads when existing ones die

### CODE REVIEW

File: lib/mini_scheduler/manager.rb [hunk 1]
```diff
@@ -352,8 +384,11 @@ def self.seq
       end
     end
 
+    @@identity_key_mutex = Mutex.new
     def identity_key
-      @identity_key ||= "_scheduler_#{hostname}:#{Process.pid}:#{self.class.seq}:#{SecureRandom.hex}"
+      @@identity_key_mutex.synchronize do
```
Comment (SamSaffron) [inline lib/mini_scheduler/manager.rb:389]: do you need to acquire the mutex if @identity_key is already defined? Just shortcut and return it?

File: lib/mini_scheduler/manager.rb [hunk 2]
```diff
@@ -163,14 +190,16 @@ def wait_till_done
...
 
-      def attempts(n)
-        n.times {
-          begin
-            yield; break
-          rescue
-            sleep Random.rand
-          end
-        }
+      def attempts(max_attempts)
```
Comment (SamSaffron) [inline lib/mini_scheduler/manager.rb:193]: yeah easier to follow for sure :+1:

File: spec/mini_scheduler/manager_spec.rb [hunk 3]
```diff
@@ -208,6 +208,58 @@ def perform
...
+      worker_thread_ids = runner.worker_thread_ids
+
+      # Now that Redis is broken, simulate the 'SuperLongJob' ending
+      worker_threads.each(&:wakeup)
+
+      # Wait until the worker dies due to the redis failure
+      wait_for(timeout: 5) { worker_threads.reject(&:alive?).count == 1 }
+
+      # Observe the broken status
+      expect(manager.schedule_info(Testing::SuperLongJob).prev_result).to eq("RUNNING")
```
Comment (SamSaffron) [inline spec/mini_scheduler/manager_spec.rb:236]: A bit hard to follow here "observe broken status" and status is RUNNING?

Comment (SamSaffron) [reply spec/mini_scheduler/manager_spec.rb:236]: I guess missing a "current status" concept?

Comment (davidtaylorhq) [reply spec/mini_scheduler/manager_spec.rb:236]: Yeah '#prev_result' returns whatever the current status is in Redis. `#result` and `#result=` are used to access/set the value which will be saved on the next `#write` call (which we're not doing in the specs).

I'll update the comment to try and make this particular case  a little clearer.

File: lib/mini_scheduler/manager.rb (hunk 1 above)
Comment (davidtaylorhq) [reply lib/mini_scheduler/manager.rb:390]: Updated

**Review by jjaffeux — APPROVED** [review]

File: spec/mini_scheduler/manager_spec.rb [hunk 4]
```diff
@@ -208,6 +208,58 @@ def perform
...
+
+      # Redis back online
+      MiniScheduler::ScheduleInfo.any_instance.unstub(:write!)
+
+      # Reschedule should not do anything straight away
+      manager.reschedule_orphans!
+      expect(manager.schedule_info(Testing::SuperLongJob).prev_result).to eq("RUNNING")
+
+      # Simultate time passing and redis keys expiring
```
Comment (pmusaraj) [inline spec/mini_scheduler/manager_spec.rb:245]:
```suggestion
      # Simulate time passing and redis keys expiring
```

File: spec/mini_scheduler/manager_spec.rb [hunk 5]
```diff
@@ -208,6 +208,58 @@ def perform
...
+      info.next_run = Time.now.to_i - 1
+      info.write!
+
+      manager.tick
+
+      wait_for do
+        manager.schedule_info(Testing::SuperLongJob).prev_result == "RUNNING"
+      end
+
+      # Simualate redis failure while job is running
```
Comment (pmusaraj) [inline spec/mini_scheduler/manager_spec.rb:222]:
```suggestion
      # Simulate redis failure while job is running
```

**Review by pmusaraj — APPROVED** [review]
Looks good to me.

File: lib/mini_scheduler/manager.rb [hunk 6]
```diff
@@ -58,12 +52,46 @@ def reschedule_orphans
...
       end
 
+      def ensure_worker_threads
+        @threads ||= []
+        @threads.delete_if { |t| !t.alive? }
+        (@manager.workers - @threads.size).times do
+          @threads << Thread.new { worker_loop }
+        end
+      rescue => ex
+        MiniScheduler.handle_job_exception(ex, message: "Scheduling manager worker thread starter")
```
[the remark this answers is not in the document]
Comment (davidtaylorhq) [reply lib/mini_scheduler/manager.rb]: Agreed - I've updated all of these to make them clearer
